src/advertisement/mod.rs

```diff
@@ -0,0 +1,189 @@
+    total_ads: i32,
+    ad_lifetime: Duration,
+    max_ads_per_topic: usize,
+    max_ads: i32,
+}
+
+impl Ads {
+    pub fn new(ad_lifetime: Duration, max_ads_per_topic: usize, max_ads: i32) -> Self {
+        Ads {
+            expirations: VecDeque::new(),
+            ads: HashMap::new(),
+            total_ads: 0,
+            ad_lifetime,
+            max_ads_per_topic,
+            max_ads,
+        }
+    }
+
+    pub fn get_ad_nodes(&self, topic: Topic) -> Result<Vec<Enr<CombinedKey>>, String> {
+        match self.ads.get(&topic) {
+            Some(topic_ads) => Ok(topic_ads
+                .into_iter()
+                .map(|ad| ad.node_record.clone())
+                .collect()),
+            None => Err("No ads for this topic".into()),
+        }
+    }
+
+    pub fn ticket_wait_time(&self, topic: Topic) -> Result<Duration, String> {
+        let now = Instant::now();
+        match self.ads.get(&topic) {
+            Some(nodes) => {
+                if nodes.len() < self.max_ads_per_topic {
+                    Ok(Duration::from_secs(0))
+                } else {
+                    match nodes.get(0) {
+                        Some(ad) => {
+                            let elapsed_time = now.saturating_duration_since(ad.insert_time);
+                            Ok(self.ad_lifetime.saturating_sub(elapsed_time))
+                        }
+                        None => {
+                            #[cfg(debug_assertions)]
+                            panic!("Panic on debug,topic key should be deleted if no ad nodes queued for it");
+                            #[cfg(not(debug_assertions))]
+                            {
+                                error!("Topic key should be deleted if no ad nodes queued for it");
+                                return Err("No nodes for topic".into());
+                            }
+                        }
+                    }
+                }
+            }
+            None => {
+                if self.total_ads < self.max_ads {
+                    Ok(Duration::from_secs(0))
+                } else {
+                    match self.expirations.get(0) {
+                        Some((fut, _)) => Ok(fut.deadline().saturating_duration_since(now)),
+                        None => {
+                            #[cfg(debug_assertions)]
+                            panic!("Panic on debug, mismatched mapping between expiration queue and total ads count");
+                            #[cfg(not(debug_assertions))]
+                            {
+                                error!("Mismatched mapping between expiration queue and total ads count");
+                                return Err("No nodes in table".into());
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn insert(&mut self, node_record: Enr<CombinedKey>, topic: Topic) -> Result<(), String> {
```

Comment:
Aha, I should use Option because of this instead of Result and return None.

Reply:
```rust
topic_hash(topic).map(|topic| {
                    self.ads.ticket_wait_time(topic).map(|wait_time| {
                        let new_ticket = Ticket::new(topic);
                        self.send_ticket_response(
                            node_address.clone(),
                            id.clone(),
                            new_ticket,
                            wait_time,
                        );

                        let _ticket = Ticket::decode(ticket).unwrap_or(Ticket::default());

                        // use wait time to see if there is any point at doing ticket validation

                        // choose which ad to reg based on ticket, for example if some node has empty ticket
                        // or is coming back, and possibly other stuff

                        match self.ads.insert(enr, topic) {
                            Ok(()) => self.send_regconfirmation_response(node_address, id, topic),
                            Err(e) => error!("{}", e),
                        }
                    }).ok();
                }).ok();
```

This is what I had, it worked when the Error returned by ```topic_hash``` and ```ticket_wait_time``` was String, but not &str. At second thought, it would have worked with &'static str right?

Reply:
Now I return Option, I think it makes more sense since I don't want to do anything if None is returned.